Regex boundaries anchors (#440)
arcus/education_modules: regular_expressions_boundaries_anchors/regular_expressions_boundaries_anchors.md

### End-of-string Anchoring 

As a reminder, the symbol for end-of-string in a regular expression is `$`.  We can use this symbol to look for specific patterns at the end of strings.  For example, we could look for specific typos that tend to appear at the end of strings: periods that follow single or double quotation marks instead of being enclosed within the quotation.  Here are some examples of acceptable grammar and typos:

**Acceptable**

* stated, "I want to withdraw from the study."
* not clear on pre-registration next steps.'

**Typos**

* reported she "did not receive the sample".
* chi-square would not be robust'. 

The pattern we're looking for is "single or double quotation followed by a period and the end of the string," or `['"]\.$`.  If you want to see this in action, you can [try it on Regex 101](https://regex101.com/r/9eMUve/1).

> 

Author: This is your suggestion, but added as a committable suggestion!

If this refresher doesn't seem familiar, you might want to consider reviewing [Demystifying Regular Expressions](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/demystifying_regular_expressions/demystifying_regular_expressions.md#1), [Regular Expressions Basics](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/regular_expressions_basics/regular_expressions_basics.md#1), and [Regular Expressions: Groups](https://liascript.github.io/course/?https://raw.githubusercontent.com/arcus/education_modules/main/regular_expressions_groups/regular_expressions_groups.md#1).

## Scope of This Module

We consider this module to be intermediate, because it assumes basic knowledge covers more advanced regular expressions topics, without delving into the advanced topic of using regular expressions within the context of code.  

Specifically, we'll address these topics:

* **Anchors**, which help you define patterns that belong in the beginning and end of a string as a whole
* **Boundaries**, which help you define patterns that belong in certain places that appear within a string

> * **Exclusions**, which allow you to look for patterns that exclude certain characters

Reviewer: This doesn't match the table of contents. I'd be okay with taking this page out entirely, but I think if it is there it should match the order and relative importance of each topic indicated by the table of contents.

> * **Exclusions**, which allow you to look for patterns that exclude certain characters

Author: @drelliche what say?  Is the above better? Or should I just delete the scope page altogether?

## Regular Expression Flags

> The regular expression checker we refer to and use for this module, [Regex 101](https://www.regex101.com), has, by default, the "global" and "multiline" **flags** set, which you can see written as "gm" after the "Regular Expression" box.  Flags in regex are parameters which can change how regular expressions are interpreted.  

Reviewer: ```suggestion
**Flags** in regex are parameters which can change how regular expressions are interpreted.  The regular expression checker we refer to and use for this module, [Regex 101](https://www.regex101.com), has, by default, the "global" and "multiline" flags set, which you can see written as "gm" after the "Regular Expression" box.   
```

switching the order of these two sentences

Flags aren't written within a regular expression, but are set **outside** of it, either by clicking on something (it might be the word "flag" or an image of a flag) or by typing a letter (like "g" or "m") or letters into your code, separate from the definition of a regular expression. Three frequently used flags are:

* **Global (g)**.  This indicates that the entire string given for matching against the pattern should be checked.  Without the "g" or "global" flag, only the first match to the pattern will be found, and the regex parser will stop looking for additional matches.

> * **Multiline (m)**.  This flag indicates that each line should be treated as its own string, instead of all the lines being considered one string.  This is important when working with start-of-string and end-of-string **anchors**, which we'll talk about later in this module.

Author: ```suggestion
* **Multiline (m)**.  This flag indicates that each line should be treated as as if it were its own string.  This is important when working with start-of-string and end-of-string **anchors**, which we'll talk about later in this module.  When the multiline flag is turned on, it means that the anchors that typically only refer to the beginning and end of a string will also be used to the beginning and end of each line. 
```
@drelliche this is regarding the multiline movie parsing example further on.

* **Ignore / Insensitive (i)**.  "Ignore" indicates that pattern matching should be case-insensitive, so that differences in capitalization are ignored.

> If you use different "flags" than we do (for example, if you're not using the default [Regex 101](https://www.regex101.com) as your regex checker), you might see an occasional difference in how the regular expression we show works in your environment.  In the regex checker you're using, try setting your flags to "global" and "multiline", and you should get identical results to what we show here in this module.  The flags usually are shown following a forward slash, appearing just after the regular expression ends:

Reviewer: ```suggestion
If you use different "flags" than we do (for example, if you're not using the default [Regex 101](https://www.regex101.com) as your regex checker), you might see an occasional difference in how the regular expression we show works in your environment.  In the regex checker you're using, try setting your flags to "global" and "multiline," and you should get identical results to what we show here in this module.  The flags usually are shown following a forward slash, appearing just after the regular expression ends:
```

> ![A regular expression checker box labeled "regular expression" includes a forward slash, the phrase "insert your regular expression here", a forward slash, and the letters "g" and "m". ](media/example1.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 400px;" --> Screenshot taken from [https://regex101.com/](https://regex101.com/)
> 
> ![A sample regular expression: `([A-Z])\w+` is enclosed within forward slashes.  After the closing forward slash, the letter "g" appears.](media/example2.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 300px;" --> Screenshot taken from [https://regexr.com/](https://regexr.com/)
> 
> ![A regular expression checker box labeled "Regular Expression" that includes two forward slashes followed by the letter "g".](media/example3.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 400px;" --> Screenshot taken from [https://www.regextester.com/](https://www.regextester.com/)

Reviewer: ```suggestion
![A regular expression checker box labeled "regular expression" includes a forward slash, the phrase "insert your regular expression here," a forward slash, and the letters g and m. ](media/example1.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 400px;" --> Screenshot taken from [https://regex101.com/](https://regex101.com/)

![A sample regular expression: `([A-Z])\w+` is enclosed within forward slashes.  After the closing forward slash, the letter "g" appears.](media/example2.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 300px;" --> Screenshot taken from [https://regexr.com/](https://regexr.com/)

![A regular expression checker box labeled "Regular Expression" that includes two forward slashes followed by the letter g.](media/example3.png)<!-- style = "border: 1px solid rgb(var(--color-highlight)); max-width: 400px;" --> Screenshot taken from [https://www.regextester.com/](https://www.regextester.com/)
```

<div class = "behind-the-scenes">
<b style="color: rgb(var(--color-highlight));">Behind the scenes</b><br>

If you do some internet searches or use a regular expression checker, you might wondering why regular expressions are often written between forward slashes.  That's because of how regular expressions are written in some programming languages like Perl and Javascript and some utilities like `sed`.  

You don't need to worry about enclosing your regular expressions in forward slashes for this module.  In fact, depending on what kind of code you write, you may never have to enclose your regular expressions in forward slashes.  Still, it's somewhat customary to see regular expressions written this way.

</div>

### Quiz: Flags

Which of the following statements are true?  Check all that apply.

[[ ]] Flags are only important if you're working in Perl.
[[ ]] Flags are written as the last letter or letters of a regular expression.
[[X]] Flags alter the rules around regular expression parsing.

> [[X]] One common flag is "global", written with a "g", which allows for multiple matches.

Reviewer: ```suggestion
[[X]] One common flag is "global," written with a g, which allows for multiple matches.
```

</div>

### Except: Another Use of the Caret 

On the previous page, we warned you about a potential pitfall:

> To correctly use `^` as the start-of-string token, you must not enclose it in square brackets.  If `^` appears inside square brackets, it will be intepreted as doing something quite different, as you are about to discover.

Let's look at three ways we can use the caret (`^`).

Start-of-string Caret
-----

If we want to indicate the start of a string, we add the caret at the beginning of our regular expression, being sure to not enclose it in square brackets.  We see an example of this in the regular expression `^(?:WT|PBMC)\-\d{5,6}`, which includes a start of string character followed by a non-capturing group with two alternatives, a hyphen, and then five to six digits.

Except-for Caret
------

We can also use the caret symbol in an "except for" clause in our regular expression.  For example, what if at a given position in a pattern, any character is possible except for a hash mark (#)?  We can say "except for" by using `^` as the first character within square brackets: `[^#]`.  

> We can add multiple characters that should not appear within the square brackets.  If any character except for  #, 5, ., and % are permitted at a given location in a string, we could express that with the regular expression list `[^#5\.%]`. For example, `[ABXYZ]` means "any of these: A, B, X, Y, or Z", while `[^ABXYZ]` means "anything that **isn't** one of these: A, B, X, Y, and Z."  

Reviewer: ```suggestion
We can add multiple characters that should not appear within the square brackets.  If any character except for  #, 5, ., and % are permitted at a given location in a string, we could express that with the regular expression list `[^#5\.%]`. For example, `[ABXYZ]` means "any of these: A, B, X, Y, or Z," while `[^ABXYZ]` means "anything that **isn't** one of these: A, B, X, Y, and Z."  
```

<div class = "important">
<b style="color: rgb(var(--color-highlight));">Important note</b><br>

> To use `^` as "except for", it has to immediately follow an opening square bracket, and be followed by the characters it is indicating aren't acceptable.  You can't, for example, say "all the digits except for 7" by typing `\d^7` or `[\d^7]`.

Reviewer: If I wanted any letter except q, would the best way to do that be [A-PR-Za-pr-z]? It might be nice to have an example here showing that you can denote "any letter except q" or something like that even though it is clunky.

Or maybe just continue the example of [0-689] or [0-68-9].

> To use `^` as "except for", it has to immediately follow an opening square bracket, and be followed by the characters it is indicating aren't acceptable.  You can't, for example, say "all the digits except for 7" by typing `\d^7` or `[\d^7]`.

Author: ```suggestion
To use `^` as "except for", it has to immediately follow an opening square bracket, and be followed by the characters it is indicating aren't acceptable.  You can't, for example, say "all the digits except for 7" by typing `\d^7` or `[\d^7]`.  One way to do that would be `[012345689]`, and another would be `[0-689]`.  

In some regular expression engines, you can do some advanced practices such as creating a **union** between two sets (such as the set of all digits and the set of all symbols except for 7), but the way to do this is different in different languages and regular expressions environments, so we won't go into detail about that here.

```

Author: @drellich what do you think about the above?

Reviewer: I think you mean the intersection of two sets (instead of union). Other than that I love this addition!

Reviewer: Oh, this is now the first introduction of the word "engine" which might benefit from a definition/disambiguation to learners new to regex.

</div>

Literal Caret
---

If you're looking for the actual symbol `^` in your text, you'll have to **escape** it by adding a backslash right before the caret: `\^`.  As a reminder, that backslash is useful for all **metacharacters** in regular expressions.  You'd do the same thing for the end-of-string character: `\$` is how you look for a literal dollar sign / money symbol.

### End-of-string Anchoring 

As a reminder, the symbol for end-of-string in a regular expression is `$`.  We can use this symbol to look for specific patterns at the end of strings.  For example, we could look for specific typos that tend to appear at the end of strings: periods that follow single or double quotation marks instead of being enclosed within the quotation.  Here are some examples of acceptable grammar and typos:

**Acceptable**

* stated, "I want to withdraw from the study."

> * not clear on pre-registration next steps.'

Reviewer: It isn't quite right to end a sentence in `.'` since single quotes only ever appear inside double quotes (see rule 7 and 8b here: https://www.grammarbook.com/punctuation/quotes.asp)

I kind of feel bad for even knowing that, let alone looking it up and being right and sharing the link....

Actually, I think that as wonderful as this is and as honored as I am that I might have drilled this punctuation rule into your head, it is actually not a great example because it is particularly un-translatable... Even other languages using the same alphabet have different punctuation rules. I'll think about another typo I can try to suggest.

Reviewer: You did make me proud!!

Author: @drelliche Can we just pretend this is a reasonable thing to search for?  I could also just search for the double quote followed by a period, but I like to complicate things by giving an example where there are two typos we're looking for.

**Typos**

* reported she "did not receive the sample".

> * chi-square would not be robust'. 

Reviewer: I actually "laughed out loud".

And then cringed as I created my own typo...

### Scenario: Start- and End-of-String Anchors

Consider these test strings, which are movie sequel titles:

* Speed 2: Cruise Control
* Teenage Mutant Ninja Turtles II: The Secret Of The Ooze
* Tomb Raider: The Cradle Of Life

You're helping a colleague who tried to write regex that would capture two groups for each movie: first, the part of the title before the colon, and second, the part of the title after the colon. 

> They used this regular expression, which uses capturing groups:  `([A-Za-z0-9\s]+):([A-Za-z0-9\s]+)`. [As we can see in Regex 101](https://regex101.com/r/3kyzIp/1), this did not segment the movie titles exactly right.  The regular expression matches the pattern twice, not three times.  It gets the first capturing group correct.  But for the second capturing group, the parser takes the second part of the first movie and keeps reading well into the second movie line, stopping only when it gets to a colon. 

Reviewer: Question about this example: the screenshot shows that the `g` and `m` flags are on, which initially made me think that Regex101 would treat each line as its own separate string, so I was confused by this behavior. But looking more closely at the Regex101 documentation I see that the `m` flag actually means that `^` and `$` match the beginning and ends of lines, not that each line is its own string.  This leads me to things:

1. Correct the phrasing describing the `m` flag on the flags page
2. Move the flags section here so that you can then give one more example with anchors: `^([A-Za-z0-9\s]+):([A-Za-z0-9\s]+)$`

Thoughts?

Author: @drelliche actually I might leave flags where it belongs and just remove this scenario portion altogether, it might be confusing and un-helpful to people.  What do you think?

Reviewer: I agree, let's take this scenario out entirely and save it for a future regex practice module.

@end

@learning_objectives  
After completion of this module, learners will be able to:

- Explain what a regular expression flag does
- Use anchors and boundaries in regular expressions
- Use boundaries in regular expressions

@end
good_first_module: false
coding_required: true
coding_level: intermediate
sequence_name: regex
previous_sequential_module: regular_expressions_groups
@sets_you_up_for

> - regular_expressions_lookaheads

Reviewer: ```suggestion

```
I don't think these modules actually depend on/set you up for each other, even though they are sequential in the sequence, which is recorded elsewhere in the metadata.

Author: agree!  adding to a batch for commit

In some regular expression environments, you can do some advanced practices such as creating an **intersection** between two sets (such as the set of all digits and the set of all symbols except for 7), but the way to do this is different in different languages and regular expressions environments, so we won't go into detail about that here.

Similarly, if you wanted to define a list with all uppercase letters except F and L, the most universal way to accomplish that is `[A-EG-KM-Z]`, but you'll sometimes see other, system-specific regular expressions that go beyond the scope of this module.

</div>

Literal Caret
---

If you're looking for the actual symbol `^` in your text, you'll have to **escape** it by adding a backslash right before the caret: `\^`.  As a reminder, that backslash is useful for all **metacharacters** in regular expressions.  You'd do the same thing for the end-of-string character: `\$` is how you look for a literal dollar sign / money symbol.

### End-of-string Anchoring 

> 

Reviewer: Ok, here is a proposed update that will satisfy my grammar prescriptivism for published materials (I try to be a descriptivist with language but punctuation is more of struggle).

```
As a reminder, the symbol for end-of-string in a regular expression is `$`.  We can use this symbol to look for specific patterns at the end of strings.  For example, we could look for specific typos that tend to appear at the end of strings: periods that follow something other than a letter or number.  Here are some examples of acceptable grammar and typos:

**Acceptable**

* This is with respect to record 117.
* Participants were not clear on pre-registration next steps.

**Typos**

* Reported she "did not receive the sample".
* A chi-squared test would not be robust . 

The pattern we're looking for is "anything that is not a letter or digit, followed by a period," or `[^\w]\.$`.  If you want to see this in action, you can [try it on Regex 101](https://regex101.com/r/4cZn0e/1).
```

Author: I like it!

But that's not what we meant!  Luckily we can fix this by being more specific.  Click the regular expressions below to see them *in situ* on Regex 101.

* We could indicate that we want two or more letters, excluding other things like spaces, with [`[A-Za-z]{2,}ome\b`](https://regex101.com/r/TdF4kr/1)
* We could use the special "word character" symbol of `\w` to indicate that letters, numbers, and underscores are what we mean: [`\w{2,}ome\b`](https://regex101.com/r/sh6CV5/1).

In both of these cases, the matches we get back are the longer words that end in "ome" -- namely, "genome," "exome," and "proteome."

<div class = "behind-the-scenes">
<b style="color: rgb(var(--color-highlight));">Behind the scenes</b><br>

There are other defined boundaries, such as the "not a word" boundary (`\B`), which isn't supported in every engine.  For now, the word boundary will probably be enough for most of the regular expressions you need to write.

</div>

## Quiz: Anchors and Boundaries

> What's the anchor character that indicates the "start of a string"?

Reviewer: ```suggestion
What's the anchor character that indicates the **start of a string**?
```

[[^]]
<script>
  let input = "@input".trim();
  input == "^";
</script>
***
<div class = "answer">

The anchor for the start of a string is the caret, or `^`, symbol.

</div>
***

> What's the anchor character that indicates the "end of a string"?

Reviewer: ```suggestion
What's the anchor character that indicates the **end of a string**?
```
